### ISSUE TITLE
ldaps - dh key too small

### ISSUE THREAD

**Issue by nicolas581** [cannatag/ldap3#249]
Hello,

I use python 3.5 and ldap3 2.1.0
The connection with ldap worked fine but now I try to use ldaps with our server. 

I got the following error: ('socket ssl wrapping error: [SSL: SSL_NEGATIVE_LENGTH] dh key too small (_ssl.c:645)',

I guess it is linked to the Weak Diffie-Hellman : https://weakdh.org/. The version of openssl used by python do not accept a too short key. Nevertheless we can't change the key on the server.

I couldn't find a way to exclude weak ciphers for the tls part in my program as mentioned for other python modules :
DEFAULT_SSL_CIPHER_LIST += 'HIGH:!DH:!aNULL'  (http://stackoverflow.com/questions/38015537/python-requests-exceptions-sslerror-dh-key-too-small)

Any help would be appreciated to exlcude or to impose the ciphers.

Nicolas

**Comment by cannatag** [comment]
This is not related to ldap3 but to the SSL module on your server.

>> Nevertheless we can't change the key on the server.

You don't have to change the key on the server, You should upgrade the SSL version,

You can also try to use an older SSL protocol. I'm not sure if this can help. Have a look at http://ldap3.readthedocs.io/ssltls.html

Try to create a Tls() object with protocol= ssl.PROTOCOL_SSLv2 or validate=ssl.CERT_NONE. Not sure if this can help.

Let me know.
Giovanni

**Comment by nicolas581** [comment]
Thanks for your answer. OpenSSL is up to date according the packages available on the server (RHEL 6.6), not sure how the certificates were generated when the ldap (opendj) was installed.
I had already tried various older protocols without success.

On this webpage : https://weakdh.org/sysadmin.html we can see how to configure popular http server to accept or not the ciphers. What I would like is to do the same on my python client program to not accept all the ciphers of our ldap server.
Perhaps I can't do that, I don't know, but changing the server ldap to exclude weak cipher (as written in the link above) is much more difficult.

In reading https://docs.python.org/3/library/ssl.html for ssl.wrap_socket you can choose the cipher allowed or with SSLContext https://docs.python.org/3/library/ssl.html#ssl.SSLContext.set_ciphers

I would like to do the same in my ldaps connection.

Regards,

Nicolas

**Comment by cannatag** [comment]
I think you should set the cipher in the ldap3/core/tls.py module BEFORE the 
```
wrapped_socket = ssl_context.wrap_socket(connection.socket, server_side=False, do_handshake_on_connect=do_handshake)
```
at line 180


you should insert:
```
ssl_context.set_ciphers('your ciphers)
```
I'm not sure about the format of the ciphers string.

**Comment by nicolas581** [comment]
I could explain longer until the end of week but I got it. I think you could update your code to be more flexible to manage this case. I don't know if I must explain here or writting an email, let me know.

To make long story short this is the _DEFAULT_CIPHER defined in ssl.py and as it is DH key allowed that's the reason.

Thank you, I'm beginner in python but I really like it and ldap3 is a good way for learning.

**Comment by cannatag** [comment]
Hi Hicolas, good to know you solved your problem. please give me more information on how you did it. I'll try to include this case in the next release of ldap3.

Bye,
Giovanni

**Comment by nicolas581** [comment]
Hello Giovanni,

Let's explain the steps even if you indicated the line :
1. Create an instance of Tls, Server and Connection
2. After some debug there is a call at strategy/base.py (l. 233)
```
try:
    self.connection.server.tls.wrap_socket(self.connection, do_handshake=True)
```
3. In wrap_socket (tls.py), the ssl_context is created :
`                ssl_context = ssl.SSLContext(self.version)`

Here we are in the __new__method if we're not in SSLv2 we can see what we load _DEFAULT_CIPHERS
```
if protocol != _SSLv2_IF_EXISTS:
    self.set_ciphers(_DEFAULT_CIPHERS)
```
which is not optimal in some cases, mainly because of bad configuration of server side.


What it would be great is to be able to put the desire default_ciphers for the application.  I think adding ciphers paramater in the __init_method of tls with None by default and set_ciphers before the call of ssl_context.wrap_socket as you mentionned.

By the way we could have different configuration depending on the application. Otherwise we must modify add the set_ciphers with some hard_coding before the call of wrap_context which is not fine if we update the packages or we can also modify the ssl.py conf but same problem even if in this case it is just a constant to modify and easier to do.

It would be great to have this paramater in a next release. I definitely prefer working with your library that using jndi with Java, thank you for your job. Let me know if you you can add it please.

Bye,

Nicolas

**Comment by cannatag** [comment]
Thanks for your detailed description.
Will work on this in the weekend. 

Bye,
Giovanni

**Comment by cannatag** [comment]
Can you ckeck code in dev? I've added a cipher=None argument in the Tls() init. It should be properly handled in different python interpreters: there are 3 cases:

recent python that use the SSLContext object: ciphers are added to the context
older python that don't use the SSLContext object: ciphers are set in the wrap_socket() method
much older python that don't allow setting ciphers at all: ciphers are not set at all

Please let me know if this works for you.

**Comment by nicolas581** [comment]
My environment is python 3.5 and an opendj ldap server and it works perfectly with the dev version. Thank you so much Giovanni.

Bye,
Nicolas

**Comment by cannatag** [comment]
Good. Will include in the next release.

**Comment by cannatag** [comment]
will be released in 2.1.1

**Comment by cannatag** [comment]
Released in 2.1.1
